### PR TITLE
XorOf IR Node implementation

### PR DESCRIPTION
**Pull request by ApexTheory**
PR for https://github.com/ergoplatform/sigma-rust/issues/356

Since `Xor` is associative, I went with a slightly different approach than the Scala implementation based on https://github.com/ScorexFoundation/sigmastate-interpreter/issues/640 by checking for an odd number of `true` values.

Let me know if you see any issues with this; instead I can rework this to use `^` bitwise XOR, perhaps in a `fold` or `reduce` operation.

In the scala version, there is also a check for `E.context.activatedScriptVersion >= 2`, do I need to take this into account?

### CODE REVIEW

File: ergotree-interpreter/src/eval/xor_of.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+use crate::eval::EvalContext;
+use crate::eval::EvalError;
+use crate::eval::Evaluable;
+use ergotree_ir::mir::xor_of::XorOf;
+
+impl Evaluable for XorOf {
+    fn eval(&self, env: &Env, ctx: &mut EvalContext) -> Result<Value, EvalError> {
+        let input_v = self.input.eval(env, ctx)?;
+        let input_v_bools = input_v.try_extract_into::<Vec<bool>>()?;
+        Ok((input_v_bools.into_iter().filter(|x| *x).count() & 1 == 1).into())
```
Comment (greenhat) [inline ergotree-interpreter/src/eval/xor_of.rs:14]: Frankly, I don't find it very readable. Although I'm biased here, since I was the one who botched the Scala implementation - https://github.com/ScorexFoundation/sigmastate-interpreter/issues/640 and now I'm blowing on the water. So let's leave it as is. :)

File: ergotree-interpreter/src/eval/xor_of.rs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    use std::rc::Rc;
+
+    proptest! {
+
+        #[test]
+        fn eval(bools in collection::vec(any::<bool>(), 10..=20)) {
+            let expr: Expr = XorOf {input: Expr::Const(bools.clone().into()).into()}.into();
+            let ctx = Rc::new(force_any_val::<Context>());
+            let res = eval_out::<bool>(&expr, ctx);
+            prop_assert_eq!(res, bools.into_iter().filter(|x| *x).count() & 1 == 1);
```
Comment (greenhat) [inline ergotree-interpreter/src/eval/xor_of.rs:37]: There is not much sense in testing it's semantic equivalence with itself. Let's use a different algo here, whatever you prefer. Also, please test on empty collection as well (extend range to 0..=20).

**Review by greenhat — CHANGES_REQUESTED** [review]
Looks good! Thank you! Please see a comment about testing.

You can ignore the check in Scala version.

File: ergotree-interpreter/src/eval/xor_of.rs (hunk 2 above)
Comment (ApexTheory) [reply ergotree-interpreter/src/eval/xor_of.rs]: I updated the test with new range (forgot to revert after doing some personal testing) and also to use a mutable val to hold the reassigned result after each XOR gate.  To your point, this solution is likely more readable than the eval function, so if you prefer maybe I swap them?

Comment (greenhat) [reply ergotree-interpreter/src/eval/xor_of.rs]: Yes, please swap it. I find XOR gate with mutable accum more readable.

**Comment by ApexTheory** [comment]
I swapped the algos between eval and test.  Eval now uses an accumulator via Iterator's `fold` method.

**Review by greenhat — APPROVED** [review]
Great! Fold looks even better here!
